### ISSUE TITLE
The problem about the protected gradient in step 1 of the algorithm

### ISSUE THREAD

**Issue by YML0823** [XuanL17/Mjolnir#1]
One question I don't understand is that in step 1 you said that you can steal the shared perturbation gradient by hiding it on the server side or waiting for the parameters to be shared during FL training. I downloaded your project and ran it, and I found a question that is, what model should this protected gradient be trained on? I tried mlp, unet (this one is my own custom), but in the end they all got loading issues. Can you please answer this question? That is, what model should be used to train this protected gradient itself, or can any model be used?

**Comment by XuanL17** [comment]
You can use your own client model to get the gradient (if you want to get protected gradient, please don’t forget to add perturbation). Better use simple CNN as a start.

You question is not so clear. I guess you miss the step to make the gradient data to 2D when put into ddpm if you straightly use my code. 

OR if you don’t know how to extract gradients, you can go to: https://github.com/XuanL17/MGIA

Please don‘t forget to star the project and cite our paper. Thanks.

**Comment by YML0823** [comment]
> 您可以使用自己的客户端模型来获取梯度（如果您想获得受保护的梯度，请不要忘记添加扰动）。最好使用简单的 CNN 作为开始。
> 
> 你的问题不太清楚。我猜如果你直接使用我的代码，你会错过将梯度数据放入 ddpm 时转换为 2D 的步骤。
> 
> 或者如果你不知道如何提取渐变，你可以访问： https: [//github.com/XuanL17/MGIA](https://github.com/XuanL17/MGIA)
> 
> 请不要忘记为该项目加注星标并引用我们的论文。谢谢。


Does the following code mean that you want to use a noise dataset to generate a disturbance gradient? This noise gradient is then loaded into DDPM for training. So if you say on github, do you say that you let the downloader provide a perturbed gradient by themselves, and this gradient is also obtained in this way? Because what I understand is that the perturbation gradient you are talking about is that after the local client has trained the model, it has added a protected gradient (e.g. LDP) with some noise perturbation before uploading the gradient, not the gradient after training with noisy data. I hope you can answer my doubts in this regard, thank you very much. Thank you!

> _**imagenet_data**_ = TrainDataset(
    "E:\COde\Master\MyDiffusionModels\MyDiffusionModelsGradient\dataset\FashionMNIST_grad2\orig",
    "E:\COde\Master\MyDiffusionModels\MyDiffusionModelsGradient\dataset\FashionMNIST_grad2\\noise")

data_loader = torch.utils.data.DataLoader(imagenet_data,
                                          batch_size=batch_size,
                                          shuffle=True,
                                          num_workers=0)

dim_mults = (1, 2, 4,)
denoise_model = Unet(
    dim=120,
    channels=channels,
    dim_mults=dim_mults
)
timesteps = 1000
schedule_name = "linear_beta_schedule"
DDPM = DiffusionModel(schedule_name=schedule_name,
                      timesteps=timesteps,
                      beta_start=0.0001,
                      beta_end=0.02,
                      denoise_model=denoise_model).to(device)

optimizer = Adam(DDPM.parameters(), lr=1e-4)
epoches = 20

Trainer = SimpleDiffusionTrainer(epoches=epoches,
                                 train_loader=data_loader,
                                 optimizer=optimizer,
                                 device=device,
                                 timesteps=timesteps)

root_path = "./saved_train_models/"
setting = "image2Size{}_channels{}_dimMults{}_timeSteps{}_scheduleName{}".format(image_size, channels, dim_mults,
                                                                                 timesteps, schedule_name)
saved_path = os.path.join(root_path, setting)
if not os.path.exists(saved_path):
    os.makedirs(saved_path)

DDPM = Trainer(DDPM, model_save_path=saved_path)

best_model_path = saved_path + '/' + 'BestModel.pth'
DDPM.load_state_dict(torch.load(best_model_path, map_location='cuda:0'))

**Comment by abccc140** [comment]
Initially, we will extract the gradient of each data within dataset and add noise commonly encountered in FL to these extracted gradients. Following this, the noised gradients will serve as the dataset for training the DDPM.

**Comment by YML0823** [comment]
> 首先，我们将提取数据集中每个数据的梯度，并将 FL 中常见的噪声添加到这些提取的梯度中。随后，带噪声的梯度将作为训练 DDPM 的数据集。

That's right, that's what I understood at the beginning as the definition of a perturbation gradient. What I understand is that the result should be a clean gradient trained on a clean dataset, and then common noise is added to that gradient to form a protected gradient. And your work in this paper is to remove the noise from this protected gradient, so what I don't understand is that there is some code in your project that looks like this（The above code is part of the code I took from your project）：
```
imagenet_data = TrainDataset(
"E:\COde\Master\MyDiffusionModels\MyDiffusionModelsGradient\dataset\FashionMNIST_grad2\orig",
"E:\COde\Master\MyDiffusionModels\MyDiffusionModelsGradient\dataset\FashionMNIST_grad2\noise")

data_loader = torch.utils.data.DataLoader(imagenet_data,

Trainer = SimpleDiffusionTrainer(epoches=epoches,
train_loader=data_loader,
optimizer=optimizer,
device=device,
timesteps=timesteps)
```

This looks like training using a noise dataset???? Or did I misunderstand it, but actually imagenet_data It's not a dataset, but a gradient obtained after training with the FashionMNIST dataset and then adding noise, so that the two gradients (FashionMNIST_grad2orig, FashionMNIST_grad2noise) are obtained separately?

**Comment by YML0823** [comment]
> > 首先，我们将提取数据集中每个数据的平均值，将 FL 中常见的噪声添加到这些提取的平均值中。然后，带噪声的平均值将作为训练 DDPM 的数据集。
> 
> 对的，这就是我一开始理解的扰动梯度的定义。我的理解是结果应该是在干净的数据集上训练出来的干净的梯度，然后把常见的噪音加到这个梯度上，形成一个受保护的梯度。而你这篇论文的工作就是把这个受保护的梯度中的噪音去除掉，所以我不明白的是，你的项目中有一部分代码是这样的（上面的代码是我从你的项目中截取的部分代码）：
> 
> ```
> imagenet_data = TrainDataset(
> "E:\COde\Master\MyDiffusionModels\MyDiffusionModelsGradient\dataset\FashionMNIST_grad2\orig",
> "E:\COde\Master\MyDiffusionModels\MyDiffusionModelsGradient\dataset\FashionMNIST_grad2\noise")
> 
> data_loader = torch.utils.data.DataLoader(imagenet_data,
> 
> Trainer = SimpleDiffusionTrainer(epoches=epoches,
> train_loader=data_loader,
> optimizer=optimizer,
> device=device,
> timesteps=timesteps)
> ```
> 
> 这个看起来像是使用噪声数据集进行训练????还是我理解错误，其实 imagenet_data 并不是数据集，而是使用 FashionMNIST 数据集训练后再添加噪声得到的梯度，这样分别得到了两个梯度（FashionMNIST_grad2orig，FashionMNIST_grad2noise）？

**Comment by XuanL17** [comment]
Hi, please refer to Fig.2 in our paper, we have conditional Mjolnir using noise gradient as a condition to help training, you can choose whether to use condition or just use clean one

**Comment by YML0823** [comment]
> 你好，请参考我们论文中的图 2，我们有条件的 Mjolnir，使用噪声梯度作为条件来帮助训练，你可以选择是否使用条件或只使用干净的条件

Yes, I have now figured out the problem. Thank you very much for your answer.
